Add documentation for topics (#4807)
dart-lang/site-www: src/tools/pub/pubspec.md

Pub.dev generates the package's thumbnail image from the first screenshot. If 
this screenshot uses animation, pub.dev uses its first frame.
 
### Topics

> Package authors can use the `topics` field to categorize their package. Topics
> can be used to assist discoverability during search with filters on pub.dev.
> Pub.dev displays the topics on the package page as well as in the search
> results.

Reviewer: **nit:** Simplify language a little. You can ignore this if you disagree.
```suggestion
Package authors can use the `topics` field to categorize their package. Topics
assist discoverability during search with filters on pub.dev.
Pub.dev displays the topics on the package page as well as in the search
results.
```

The field consists of a list of names. For example:

```yaml
topics:
  - network
  - http
```

> Pub.dev limits topics to the following specifications:
> 
> - At most 5 topics per package.
> - The length of the topic name should be 2-32 characters.
> - The topic name should consist of lowercase alphanumeric characters or dash
>   (but no double dash), starting with a-z and ending with a-z or 0-9.

Reviewer: **issue:** This needs parallel structure, per [GSG](https://developers.google.com/style/lists#parallel-syntax)

```suggestion
Pub.dev requires topics to follow these specifications:

- Use no more than 5 topics per package.
- Set the length of the topic name to between 2 and 32 characters.
- Write the topic name following these requirements:
  - Use only lowercase alphanumeric characters or hyphens (`a-z`, `0-9`, `-`).
  - Don't use two consecutive hyphens.
  - Start the name with lowercase alphabet characters (`a-z`).
  - End with alphanumeric characters (`a-z` or `0-9`).
```

Reviewer 2: > - Set the length of the topic name to between 2 and 32 characters.

I don't think it makes sense to write that.

---------------------

We could perhaps do:
```suggestion
Pub.dev limits topics for a package as follows:
- At most 5 topics per package.
- Each topic name must:
  - Use only lowercase alphanumeric characters or hyphens (`a-z`, `0-9`, `-`).
  - Don't use two consecutive hyphens.
  - Start the name with lowercase alphabet characters (`a-z`).
  - End with alphanumeric characters (`a-z` or `0-9`).
  - Be between 2 and 32 characters long.
```

Author: Updated the bullet list.

Pub.dev generates the package's thumbnail image from the first screenshot. If 
this screenshot uses animation, pub.dev uses its first frame.
 
### Topics

Package authors can use the `topics` field to categorize their package. Topics
can be used to assist discoverability during search with filters on pub.dev.
Pub.dev displays the topics on the package page as well as in the search
results.

The field consists of a list of names. For example:

```yaml
topics:
  - network
  - http
```

Pub.dev requires topics to follow these specifications:

- Tag each package with at most 5 topics.
- Write the topic name following these requirements:
  - Use between 2 and 32 characters.
  - Use only lowercase alphanumeric characters or hyphens (`a-z`, `0-9`, `-`).
  - Don't use two consecutive hyphens (`--`).
  - Start the name with lowercase alphabet characters (`a-z`).

>   - End with alphanumeric characters (`a-z` or `0-9`).

Reviewer 2: This is not the specification I was promised on the line above.

Pub.dev generates the package's thumbnail image from the first screenshot. If 
this screenshot uses animation, pub.dev uses its first frame.
 
### Topics

Package authors can use the `topics` field to categorize their package. Topics
can be used to assist discoverability during search with filters on pub.dev.
Pub.dev displays the topics on the package page as well as in the search
results.

The field consists of a list of names. For example:

```yaml
topics:
  - network
  - http
```

Pub.dev limits topics to the following specifications:

- At most 5 topics per package.
- The length of the topic name should be 2-32 characters.
- The topic name should consist of lowercase alphanumeric characters or dash
  (but no double dash), starting with a-z and ending with a-z or 0-9.

> 

Reviewer: **suggestion:** Add a heading for "Best Practices".
```markdown
#### Best Practices
```

Author: done

Reviewer 2: We don't do this for other subjects on the page.

We could put it in an "Note" block, but I think that's overkill.

> When choosing topics it is good practice to look at
> [existing topics]({{site.pub}}/topics) to see if there are established topics
> that match your package.

Reviewer: **issue:** Write in active voice per [GSG](https://developers.google.com/style/voice).

```suggestion
When choosing topics, first look at
[existing topics]({{site.pub}}/topics) to check if established topics
exist that match your package.
```

Author: done

Reviewer 2: We should perhaps consider being less specific here.

Using a word like "may" is or say "may wish to", is perhaps a good idea.

There may be established topics that match your package, but which you don't want to use.
They may be too generic (then they still technically match your package), but it may be that you want to establish a new more specific topic -- that's fine, apply human judgement.

> Another good idea, if you have a number of related packages, for instance
> belonging to the same framework, it is good practice to tag of all these
> packages with the same topic, e.g. the name of the framework.

Reviewer: **issues:** Rephrase to make more concise and readable. 
Write in active voice per [GSG](https://developers.google.com/style/voice).

```suggestion
If you have a number of related packages,
tag of all these packages with the same topic.
Such packages could belong to the same framework
and be tagged with the name of the framework.
```

Reviewer 2: > If you have a number of related packages,
> tag of all these packages with the same topic.

This changes the sentence from a recommendation to an instruction.

I think we should give advice, not instructions.
Package authors are writing code, this is complicated and we rely on them using their best judgement when it comes to presenting their package.

Hence, I would suggest we include words such as "may" and "should" or "consider", to explain to readers whether something is a suggestion or requirement.

> Such packages could belong to the same framework
> and be tagged with the name of the framework.

This just reads wrong to me.

Pub.dev generates the package's thumbnail image from the first screenshot. If 
this screenshot uses animation, pub.dev uses its first frame.
 
### Topics

Package authors can use the `topics` field to categorize their package. Topics
can be used to assist discoverability during search with filters on pub.dev.
Pub.dev displays the topics on the package page as well as in the search
results.

The field consists of a list of names. For example:

```yaml
topics:
  - network
  - http
```

Pub.dev limits topics to the following specifications:

- At most 5 topics per package.
- The length of the topic name should be 2-32 characters.
- The topic name should consist of lowercase alphanumeric characters or dash
  (but no double dash), starting with a-z and ending with a-z or 0-9.

> #### Best Practices
> 
> When choosing topics, first look at [existing topics]({{site.pub}}/topics) to
> check if established topics exist that match your package.
> 
> If you have a number of related packages, tag of all these packages with the
> same topic. Such packages could belong to the same framework and be tagged with
> the name of the framework.
> 

Reviewer 2: ```suggestion
When choosing topics it is good practice to look at
[existing topics]({{site.pub}}/topics) to see if there are established topics
that match your package.

Another good idea, if you have a number of related packages, for instance
belonging to the same framework, it is good practice to tag of all these
packages with the same topic, e.g. the name of the framework.
```
I suggest we go back to the original proposal.

--------------

```
#### Best Practices
```
We don't do this other places on this page.

> When choosing topics, first look at [existing topics]({{site.pub}}/topics) to
> check if established topics exist that match your package.

This is an instruction and not a recommendation or suggestion.
The original intent here is to suggest that users consider existing topics.

> If you have a number of related packages, tag of all these packages with the
> same topic. Such packages could belong to the same framework and be tagged with
> the name of the framework.

Again, we should suggest not instruct. Also the part about frameworks does not make sense to me.

Reviewer 2: If we wanted a different text I'd suggest:

```suggestion
When publishing a package consider if [existing topics]({{site.pub}}/topics)
are relevant. Tagging with existing topics helps users discover your package.

If you have a number of related packages, for example packages making up a framework,
consider establishing a new topic linking these packages together. This makes it easy
to discover related packages, and allows other package authors to tag their packages
as being relevant for your framework. For example, check out
[`topic:shelf`](https://pub.dev/packages?q=topic%3Ashelf).
```

We can also omit these lines entirely and just put the recommendations in `publishing.md`.

Reviewer 3: \cc @atsansone and @szakarias for thoughts on content and text

I think you are right though that these extensive details don't belong here and could be a section in [Publishing](https://dart.dev/tools/pub/publishing).
